### PR TITLE
mgr/dashboard: Fix for incorrect validation in rgw user form

### PR DESCRIPTION
**Pull request by nizamial09**
The rgw users create form doesnt validate the username correctly if the username is a tenated one. For eg. Consider there is a user called tenate$sample. Now I am trying to create another user and I entered tenate$sample as username. But it doesn't async validate the username as existing. Instead it just shows the green tick and once the submit button is clicked it'll show the user as existing.

users

**BEFORE**
validation
**AFTER**
aftervalidation

Fixes: https://tracker.ceph.com/issues/48907
Signed-off-by: Nizamudeen A <nia@redhat.com>

## Checklist
- [x] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by nizamial09** [comment]
jenkins test make check

**Review by epuertat — APPROVED** [review]

**Review by votdev — APPROVED** [review]
LGTM

**Review by s0nea — APPROVED** [review]

**Review by votdev — CHANGES_REQUESTED** [review]
You need to adapt https://github.com/ceph/ceph/blob/master/src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-user-form/rgw-user-form.component.spec.ts#L157.

The new validator code LGTM.

**Comment by votdev** [comment]
Please append the following patch to adapt the unit tests to your changes:
```
diff --git a/src/pybind/mgr/dashboard/frontend/src/app/shared/api/rgw-user.service.spec.ts b/src/pybind/mgr/dashboard/frontend/src/app/shared/api/rgw-user.service.spec.ts
--- a/src/pybind/mgr/dashboard/frontend/src/app/shared/api/rgw-user.service.spec.ts	(revision 06b2f05c7ae09f99ac6dd88e059e33765d8b743f)
+++ b/src/pybind/mgr/dashboard/frontend/src/app/shared/api/rgw-user.service.spec.ts	(date 1611245419257)
@@ -1,7 +1,7 @@
 import { HttpClientTestingModule, HttpTestingController } from '@angular/common/http/testing';
 import { TestBed } from '@angular/core/testing';
 
-import { of as observableOf } from 'rxjs';
+import { of as observableOf, throwError } from 'rxjs';
 
 import { configureTestBed } from '~/testing/unit-test-helper';
 import { RgwUserService } from './rgw-user.service';
@@ -138,21 +138,19 @@
     expect(req.request.method).toBe('DELETE');
   });
 
-  it('should call exists with an existent uid', () => {
-    spyOn(service, 'enumerate').and.returnValue(observableOf(['foo', 'bar']));
-    let result;
+  it('should call exists with an existent uid', (done) => {
+    spyOn(service, 'get').and.returnValue(observableOf({}));
     service.exists('foo').subscribe((res) => {
-      result = res;
+      expect(res).toBe(true);
+      done();
     });
-    expect(result).toBe(true);
   });
 
-  it('should call exists with a non existent uid', () => {
-    spyOn(service, 'enumerate').and.returnValue(observableOf(['foo', 'bar']));
-    let result;
+  it('should call exists with a non existent uid', (done) => {
+    spyOn(service, 'get').and.returnValue(throwError('bar'));
     service.exists('baz').subscribe((res) => {
-      result = res;
+      expect(res).toBe(false);
+      done();
     });
-    expect(result).toBe(false);
   });
 });
===================================================================
diff --git a/src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-user-form/rgw-user-form.component.spec.ts b/src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-user-form/rgw-user-form.component.spec.ts
--- a/src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-user-form/rgw-user-form.component.spec.ts	(revision 06b2f05c7ae09f99ac6dd88e059e33765d8b743f)
+++ b/src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-user-form/rgw-user-form.component.spec.ts	(date 1611245728538)
@@ -7,7 +7,7 @@
 import { NgbTooltipModule } from '@ng-bootstrap/ng-bootstrap';
 import { NgxPipeFunctionModule } from 'ngx-pipe-function';
 import { ToastrModule } from 'ngx-toastr';
-import { of as observableOf } from 'rxjs';
+import { of as observableOf, throwError } from 'rxjs';
 
 import { RgwUserService } from '~/app/shared/api/rgw-user.service';
 import { NotificationType } from '~/app/shared/enum/notification-type.enum';
@@ -155,21 +155,19 @@
   });
 
   describe('username validation', () => {
-    beforeEach(() => {
-      spyOn(rgwUserService, 'enumerate').and.returnValue(observableOf(['abc', 'xyz']));
-    });
-
     it('should validate that username is required', () => {
       formHelper.expectErrorChange('uid', '', 'required', true);
     });
 
     it('should validate that username is valid', fakeAsync(() => {
+      spyOn(rgwUserService, 'get').and.returnValue(throwError('foo'));
       formHelper.setValue('uid', 'ab', true);
       tick(500);
       formHelper.expectValid('uid');
     }));
 
     it('should validate that username is invalid', fakeAsync(() => {
+      spyOn(rgwUserService, 'get').and.returnValue(observableOf({}));
       formHelper.setValue('uid', 'abc', true);
       tick(500);
       formHelper.expectError('uid', 'notUnique');
===================================================================
diff --git a/src/pybind/mgr/dashboard/frontend/src/app/shared/api/rgw-user.service.ts b/src/pybind/mgr/dashboard/frontend/src/app/shared/api/rgw-user.service.ts
--- a/src/pybind/mgr/dashboard/frontend/src/app/shared/api/rgw-user.service.ts	(revision 06b2f05c7ae09f99ac6dd88e059e33765d8b743f)
+++ b/src/pybind/mgr/dashboard/frontend/src/app/shared/api/rgw-user.service.ts	(date 1611245443841)
@@ -134,7 +134,9 @@
     return this.get(uid).pipe(
       mapTo(true),
       catchError((error) => {
-        error.preventDefault();
+        if (_.isFunction(error.preventDefault)) {
+          error.preventDefault();
+        }
         return observableOf(false);
       })
     );
```

File: src/pybind/mgr/dashboard/frontend/src/app/shared/api/rgw-user.service.ts [hunk 1]
```diff
@@ -131,10 +131,11 @@ export class RgwUserService {
    * @return {Observable<boolean>}
    */
   exists(uid: string): Observable<boolean> {
-    return this.enumerate().pipe(
-      mergeMap((resp: string[]) => {
-        const index = _.indexOf(resp, uid);
-        return observableOf(-1 !== index);
+    return this.get(uid).pipe(
+      mapTo(true),
+      catchError((error) => {
```
Comment (votdev) [inline src/pybind/mgr/dashboard/frontend/src/app/shared/api/rgw-user.service.ts:136]: If possibe please define the type (HttpEvent????), so TypeScript is aware of the type.
```suggestion
      catchError((error: XXXX) => {
```

**Review by votdev — CHANGES_REQUESTED** [review]

File: src/pybind/mgr/dashboard/frontend/src/app/shared/api/rgw-user.service.ts (hunk 1 above)
Comment (nizamial09) [reply src/pybind/mgr/dashboard/frontend/src/app/shared/api/rgw-user.service.ts]: It was actually an Event I think. @votdev

**Comment by nizamial09** [comment]
@votdev Thank you for helping me with the test. It was a tough one for me.

**Review by votdev — APPROVED** [review]

**Pull request merged**